### PR TITLE
Add Collisions plugin and Collider components

### PR DESCRIPTION
**Pull request by luishfonseca**
Depends on #418

### CODE REVIEW

File: engine/include/cubos/engine/collisions/colliders/box.hpp [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+namespace cubos::engine
+{
+    struct [[cubos::component("cubos/box_collider", VecStorage)]] BoxCollider
+    {
+        glm::mat4 transform;
+        cubos::core::geom::Box shape;
+        float margin;
+    };
+}
```
Comment (luishfonseca) [inline engine/include/cubos/engine/collisions/colliders/box.hpp:13]:
```suggestion
}

```

File: engine/include/cubos/engine/collisions/colliders/capsule.hpp [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+#include <cubos/core/geom/capsule.hpp>
+
+namespace cubos::engine
+{
+    struct [[cubos::component("cubos/capsule_collider", VecStorage)]] CapsuleCollider
+    {
+        glm::mat4 transform;
+        cubos::core::geom::Capsule shape;
+    };
+}
```
Comment (luishfonseca) [inline engine/include/cubos/engine/collisions/colliders/capsule.hpp:12]:
```suggestion
}

```

File: engine/include/cubos/engine/collisions/colliders/plane.hpp [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+#include <cubos/core/geom/plane.hpp>
+
+namespace cubos::engine
+{
+    struct [[cubos::component("cubos/plane_collider", VecStorage)]] PlaneCollider
+    {
+        glm::vec3 offset;
+        cubos::core::geom::Plane shape;
+    };
+}
```
Comment (luishfonseca) [inline engine/include/cubos/engine/collisions/colliders/plane.hpp:12]:
```suggestion
}

```

File: engine/include/cubos/engine/collisions/colliders/simplex.hpp [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+namespace cubos::engine
+{
+    struct [[cubos::component("cubos/simplex", VecStorage)]] SimplexCollider
+    {
+        glm::vec3 offset;
+        cubos::core::geom::Simplex shape;
+        float margin;
+    };
+}
```
Comment (luishfonseca) [inline engine/include/cubos/engine/collisions/colliders/simplex.hpp:13]:
```suggestion
}

```

File: engine/include/cubos/engine/collisions/plugin.hpp [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    /// - `cubos.collisions.broad.pairs`: broad phase collision pairs have been found.
+    /// - `cubos.collisions.broad`: broad phase has ended.
+    /// - `cubos.collisions`: collisions have been resolved.
+    ///
+    /// Dependencies:
+    /// - `transformPlugin`
+    ///
+    /// @param cubos CUBOS. main class.
+    void collisionsPlugin(Cubos& cubos);
+} // namespace cubos::engine
```
Comment (luishfonseca) [inline engine/include/cubos/engine/collisions/plugin.hpp:42]:
```suggestion
} // namespace cubos::engine

```

File: engine/src/cubos/engine/collisions/plugin.cpp [hunk 6]
```diff
@@ -0,0 +1,9 @@
+#include <cubos/engine/collisions/plugin.hpp>
+
+void cubos::engine::collisionsPlugin(Cubos& cubos)
+{
+    cubos.addComponent<BoxCollider>();
+    cubos.addComponent<SimplexCollider>();
+    cubos.addComponent<CapsuleCollider>();
+    cubos.addComponent<PlaneCollider>();
+}
```
Comment (luishfonseca) [inline engine/src/cubos/engine/collisions/plugin.cpp:9]:
```suggestion
}

```

**Review by luishfonseca — COMMENTED** [review]
missing newlines at eof

File: engine/include/cubos/engine/collisions/colliders/simplex.hpp [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+
+#include <cubos/core/geom/simplex.hpp>
+
+namespace cubos::engine
+{
+    struct [[cubos::component("cubos/simplex", VecStorage)]] SimplexCollider
+    {
+        glm::vec3 offset;
+        cubos::core::geom::Simplex shape;
+        float margin;
```
Comment (RiscadoA) [inline engine/include/cubos/engine/collisions/colliders/simplex.hpp:11]: This should be well documented.
Also, are we using floats or doubles?

**Review by RiscadoA — CHANGES_REQUESTED** [review]

File: engine/include/cubos/engine/collisions/colliders/simplex.hpp (hunk 7 above)
Comment (luishfonseca) [reply engine/include/cubos/engine/collisions/colliders/simplex.hpp]: I don't think we need double precision for collisions. Here speed is a requirement more important than accuracy.

File: core/tests/geom/capsule.cpp [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+    SUBCASE("height")
+    {
+        CHECK_EQ(capsule.height(), 2 * r + l);
+    }
+
+    SUBCASE("sphere")
+    {
+        auto sphere = Capsule::sphere(1.0f);
+        CHECK_EQ(sphere.height(), 2 * r);
+    }
```
Comment (RiscadoA) [inline core/tests/geom/capsule.cpp:22]:
```suggestion
    auto r = 1.0f;
    
    SUBCASE("capsule")
    {
        auto l = 2.0f;
        Capsule capsule{r, l};
        CHECK_EQ(capsule.height(), 2 * r + l);
    }

    SUBCASE("sphere")
    {
        auto sphere = Capsule::sphere(r);
        CHECK_EQ(sphere.height(), 2 * r);
    }
```

**Review by RiscadoA — CHANGES_REQUESTED** [review]

File: engine/include/cubos/engine/collisions/colliders/box.hpp [hunk 9]
```diff
@@ -0,0 +1,17 @@
+/// @file
+/// Contains the class for the Box Collider component.
```
Comment (RiscadoA) [inline engine/include/cubos/engine/collisions/colliders/box.hpp:2]:
```suggestion
/// @brief Contains the class for the Box Collider component.
```

**Review by RiscadoA — CHANGES_REQUESTED** [review]

**Review by RiscadoA — DISMISSED** [review]

**Review by RiscadoA — APPROVED** [review]

File: core/include/cubos/core/geom/box.hpp [hunk 10]
```diff
@@ -5,12 +5,15 @@
...
+#include <cubos/core/data/deserializer.hpp>
+#include <cubos/core/data/serializer.hpp>
+
 namespace cubos::core::geom
 {
     /// @brief Box shape.
     struct Box
     {
-        const glm::vec3 halfSize; ///< The half size of the box.
+        glm::vec3 halfSize{0.0f}; ///< The half size of the box.
```
Comment (RiscadoA) [inline core/include/cubos/core/geom/box.hpp:16]: Why not default to a unit cube? Would be nice for the editor.

File: core/include/cubos/core/geom/capsule.hpp [hunk 11]
```diff
@@ -3,13 +3,16 @@
...
+
 namespace cubos::core::geom
 {
     /// @brief Capsule shape, which can also represent a sphere.
     struct Capsule
     {
-        const float radius; ///< The radius of the capsule.
-        const float length; ///< The length of the capsule.
+        float radius; ///< The radius of the capsule.
+        float length; ///< The length of the capsule.
```
Comment (RiscadoA) [inline core/include/cubos/core/geom/capsule.hpp:15]: Default to a unit sphere perhaps?

File: core/include/cubos/core/geom/plane.hpp [hunk 12]
```diff
@@ -5,11 +5,38 @@
...
+#include <cubos/core/data/deserializer.hpp>
+#include <cubos/core/data/serializer.hpp>
+
 namespace cubos::core::geom
 {
     /// @brief Plane shape.
     struct Plane
     {
-        const glm::vec3 normal; ///< The normal of the plane.
+        glm::vec3 normal{0.0f}; ///< The normal of the plane.
```
Comment (RiscadoA) [inline core/include/cubos/core/geom/plane.hpp:16]: Should not default to an invalid value - set it to face Y+ by default.

**Review by RiscadoA — CHANGES_REQUESTED** [review]

**Review by RiscadoA — APPROVED** [review]

**Review by RiscadoA — APPROVED** [review]

**Pull request merged**